src/Streamly/Internal/LZ4.hs

```diff
@@ -0,0 +1,404 @@
+                       compressedSize <-
+                           peek (castPtr (b `plusPtr` 4) :: Ptr Word32)
+                       putDivider
+                       putStrLn $ "Index        : " ++ show i
+                       putStrLn $ "Length       : " ++ show len
+                       putStrLn $ "Compressed   : " ++ show compressedSize
+                       putStrLn $ "Decompressed : " ++ show decompressedSize
+
+    {-# INLINE_LATE step #-}
+    step gst (i, st) = do
+        r <- step0 gst st
+        case r of
+            Stream.Yield arr st1 -> do
+                liftIO $ debugger i arr
+                return $ Stream.Yield arr (i + 1, st1)
+            Stream.Skip st1 -> return $ Stream.Skip (i, st1)
+            Stream.Stop -> do
+                liftIO putDivider
+                return Stream.Stop
+
+-- | A simple combinator that prints the index, length, compressed size and
+-- decompressed size of each array element to standard output.
+--
+-- This only works on stream of resized arrays.
+--
+{-# INLINE debug #-}
+debug :: MonadIO m
+    => SerialT m (Array.Array Word8) -> SerialT m (Array.Array Word8)
+debug m = Stream.fromStreamD (debugD (Stream.toStreamD m))
+
+--------------------------------------------------------------------------------
+-- Primitives
+--------------------------------------------------------------------------------
+
+-- Having NOINLINE here does not effect the performance a lot. Every
+-- iteration of the loop is a little slower (< 1us) but the entire loop
+-- fuses.
+-- On a stream with 404739 elements of 10 bytes each,
+-- With NOINLINE: 96.14 ms
+-- With INLINE:   81.07 ms
+--
+-- With INLINE statement and the usage of fusion-plugin results in an
+-- enormous code size when used with other combinators.
+-- | Primitive function to compress a chunk of Word8.
+{-# NOINLINE compressChunk #-}
+compressChunk ::
+       Int -> Ptr C_LZ4Stream -> Array.Array Word8 -> IO (Array.Array Word8)
+compressChunk i ctx arr = do
+    Array.asPtr arr
+        $ \src -> do
+              let srcLen = fromIntegral $ Array.byteLength arr
+              maxCLen <- c_compressBound srcLen
+              -- allocate compressed block with 8 byte header.  First 4
+              -- bytes of the header store the length of the uncompressed
+              -- data and the next 4 bytes store the length of the
+              -- compressed data.
+              (MArray.Array fptr dstBegin dstMax) <-
+                  MArray.newArray (fromIntegral maxCLen + 8)
+              let hdrSrcLen :: Ptr Word32 = castPtr dstBegin
+                  hdrCompLen :: Ptr Word32 = dstBegin `plusPtr` 4
+                  compData = dstBegin `plusPtr` 8
+              compLen <-
+                  c_compressFastContinue
+                      ctx src compData srcLen maxCLen (fromIntegral i)
+              poke hdrSrcLen (fromIntegral srcLen)
+              poke hdrCompLen (fromIntegral compLen)
```

Comment:
We should use a specific byte ordering instead of the machine byte ordering otherwise the code will not work across machines with different byte ordering. We can raise an issue for this and address it separately.